File: CS50xIntroToComputerScience/Week1/ProblemSet/credit/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>

bool checksum(long card, int length);
void categorize(long card, int length);

int main(void)
{
    long card = get_long("Input card number:\n");
    long temp_card = card;

    int length = 0;
    while (temp_card)
    {
        temp_card /= 10;
        length++;
    }

    if (checksum(card, length))
    {
        categorize(card, length);
    }
    else
    {
        printf("INVALID\n");
    }
}

bool checksum(long card, int length)
{
    int sum = 0;

    for (int i = 1; i <= length; i++)
    {
        int digit = card % 10;
        card /= 10;

        if (i % 2 == 1)
        {
            sum += digit;
        }
        else
        {
            int doubled = digit * 2;
            sum += (doubled / 10) + (doubled % 10);
        }
    }
    return sum % 10 == 0;
}

void categorize(long card, int length)
{
    int first_two = 0;

    while (card >= 100)
    {
        card /= 10;
    }
    first_two = card;

    if ((length == 13 || length == 16) && (first_two / 10 == 4))
    {
        printf("VISA\n");
    }
    else if (length == 15 && (first_two == 34 || first_two == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (first_two >= 51 && first_two <= 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week6/ProblemSet/sentimental-credit/credit.py
import re


def main():
    pattern = r"^\d+$"
    cc = input("Number: ").strip()

    if not re.fullmatch(pattern, cc):
        print("INVALID\n")
        return

    length = len(cc)

    if checksum(cc, length):
        categorize(cc, length)
    else:
        print("INVALID\n")


def checksum(card, length):
    sum = 0
    card_number = [int(digit) for digit in card]

    for i in range(length):
        if (length - i) % 2 == 0:
            doubled = card_number[i] * 2
            sum += doubled // 10 + doubled % 10
        else:
            sum += card_number[i]

    return sum % 10 == 0


def categorize(card, length):
    first_digit = int(card[0])
    first_two_digits = int(card[:2])

    if (length == 13 or length == 16) and first_digit == 4:
        print("VISA\n")
    elif length == 15 and (first_two_digits == 34 or first_two_digits == 37):
        print("AMEX\n")
    elif length == 16 and 51 <= first_two_digits <= 55:
        print("MASTERCARD\n")
    else:
        print("INVALID\n")


if __name__ == "__main__":
    main()
